articles/execute_process_refactor.md

```diff
@@ -0,0 +1,596 @@
+Different packages may be installed, different environment variables may be defined, and so forth.
+An subclass implementation of the proposed `launch.descriptions.Executable` class may choose to override the base `apply_context` method to implement additional layer(s) of substitutions appropriate to its destination contexts, such as remote installation paths or environment variables.
+Additionally, the available support for concepts such as signals and PIDs may be different; to that end, it is likely that remote execution of processes will require sibling classes to those described below, rather than direct subclasses.
+
+## Proposed New Classes
+
+### launch.descriptions.Executable
+
+This class would represent only the information required to define an entity which might be run.
+It would not include execution-time details, nor provide monitoring of the process after launch. 
+
+#### Constructor
+
+The constructor parameters would be pulled directly from the current `launch.actions.ExecuteProcess` class.
+
+| Argument        | Description                                                  |
+| --------------- | ------------------------------------------------------------ |
+| cmd             | A list where the first item is the executable and the rest are arguments to the executable, each item may be a string or a list of strings and `Substitution`s to be resolved at runtime |
+| name            | The label used to represent the process, as a string or a `Substitution` to be resolved at runtime, defaults to the basename of the executable |
+| cwd             | The directory in which to run the executable                 |
+| env             | Dictionary of environment variables to be used, starting from a clean environment. If `None`, the current environment is used. |
+| additional\_env | Dictionary of environment variables to be added. If `env` was `None`, they are added to the current environment. If not, `env` is updated with `additional_env`. |
+
+#### Properties
+
+Accessors would be provided for the various constructor parameters: `cmd`, `name`, `cwd`, `env`, and `additional_env`.
+
+#### Methods
+
+In the current classes, substitutions are handled through internal method calls.
+To support better isolation of class functionality, substitutions will be performed in this class, but exposed from a public method which can be overridden by subclasses to perform additional appropriate substitutions.
+Substituted versions of the various parameters will be stored in a local dictionary, similar to the current `__process_event_args`.
+
+| Name          | Description                                                  |
+| ------------- | ------------------------------------------------------------ |
+| apply_context | Takes a given LaunchContext and performs actions necessary to prepare to execute the defined command in that context. This will primarily consist of expanding various substitutions. |
```

Comment:
@roger-strain please, could you remind me why we have `apply_context` _and_ `prepare` (as opposed to just `prepare`)? If it's for (semantic) consistency's sake, I can see several variations of `prepare` already.

Reply:
@hidmic I don't know if there's a specific reason for both. It might have evolved from building from both ends, if you will. If you'd rather simplify that down to just one or the other, I can definitely look at doing that.

Reply:
The latter sounds good.